### PR TITLE
Lock access to native kafka

### PR DESCRIPTION
**Pull request by thijsc**
Put all access on the native kafka behind locks, so that we can exit cleanly.

### CODE REVIEW

**Comment by mensfeld** [comment]
ref https://github.com/appsignal/rdkafka-ruby/issues/234

you want me to crash-test it? :rofl:

**Comment by thijsc** [comment]
> you want me to crash-test it? 🤣

Yes please :-)

**Comment by mensfeld** [comment]
@thijsc FYI did not forget about it. Should be able to do this at the end of this week for both waterdrop and karafka

**Comment by mensfeld** [comment]
@thijsc one more thing that is worrying (not related to this code but may be worth fixing altogether). Sometimes admin actions callback is not called back :/ no idea yet why it's happening, though.

**Comment by mensfeld** [comment]
Sorry, I was occupied running this instead first: https://github.com/appsignal/rdkafka-ruby/pull/242 - for this I can confirm that producer, consumer and admin work as expected without any issues.

For this (238) I will be running this today and tomorrow and will post you.

**Comment by mensfeld** [comment]
Ok running stress tests on producer now...

**Comment by mensfeld** [comment]
@thijsc good news and bad news. Let me sum it up with a meme:

indeks

**Comment by mensfeld** [comment]
gdb claims it hangs on close:

```
from /mnt/software/Karafka/rdkafka-ruby/lib/rdkafka/producer.rb:45:in `close'
from /mnt/software/Karafka/rdkafka-ruby/lib/rdkafka/native_kafka.rb:72:in `close'
from /mnt/software/Karafka/rdkafka-ruby/lib/rdkafka/native_kafka.rb:72:in `join'
```

for not-closed producers.

**Comment by mensfeld** [comment]
ok, giving it a spin from the producer pespective...

**Comment by mensfeld** [comment]
@thijsc question / remark:

The newly introduced `#flush` has a comment like this:

# Call this before closing a producer to ensure delivery of all messages.

but this contradicts the librdkafka docs themselves and usage of `rd_kafka_outq_len`:

 > @brief Wait until all outstanding produce requests, et.al, are completed.
 >        This should typically be done prior to destroying a producer instance
 >        to make sure all queued and in-flight produce requests are completed
 >        before terminating.

Using both is a huge problem for two reasons:
- you cannot flush from a closed producer after the close
- if you flush before the close (including states) there is a small gap where new messages can be dispatched despite the flush because producer is not technically closed yet. This will be a problem in a multi-thread env.

So we can do two things:
- ditch `rd_kafka_outq_len` (not recommended)
- update flush docs to indicate, that close producer flush happens and that this is NOT needed (because it is not) but CAN be used for periodic flushes to dispatch messages faster, et (recommended)

WDYT? This is a big **blocker** for me, because it's either one or the other.

**Comment by mensfeld** [comment]
Actually my PR that got merged could then be reverted if we would document that flush has a potential race condition on close but it is irrelevant because of `rd_kafka_outq_len` usage.

**Comment by mensfeld** [comment]
With my PR reverted AND flush before initialization of close + `rd_kafka_outq_len` on wait, producer is **stable**. No crashes there anymore.

Now for the consumer...

**Comment by mensfeld** [comment]
@thijsc I don't have rights to restart semaphore buuuuut it's :green_heart: 

What I did:
- reorder the shutdown process to ensure that queue is empty and closed before we destroy the inner kafka
- reorder to prevent deadlock
- move kafka start in docker higher up so it fully starts while things are bundled and compiled - I had the same problem with my regressions. Unless Kafka starts and gets a bit of time to do housekeeping, fast topic creation can hang it during warmup.

- My integration producer specs now fully pass under load
- My integration suite also. No segfaults.

**Review by nijikon — APPROVED** [review]
LGTM 🚀

**Comment by mensfeld** [comment]
Ok one more challenge I see now for finalizers: trap context. Will fix.

**Comment by thijsc** [comment]
You're on a roll!

**Comment by mensfeld** [comment]
In total there were 14 dangling rdkafka objects, some with finalizers and some without (direct references of inner).

A potential multi-thread race condition was also fixed.

Another crash was caused by leaked consumes being present during fork and attempt to close them when fork was finishing. COW won't work with open sockets.

File: spec/rdkafka/consumer_spec.rb [hunk 1]
```diff
@@ -738,7 +738,8 @@ def new_message
       #
       # This is, in effect, an integration test and the subsequent specs are
       # unit tests.
-      create_topic_handle = rdkafka_config.admin.create_topic(topic_name, 1, 1)
+      admin = dkafka_config.admin
```
Comment (nijikon) [inline spec/rdkafka/consumer_spec.rb:741]: `rdkafka_config`

Comment (mensfeld) [reply spec/rdkafka/consumer_spec.rb]: Fixed. Thanks

**Comment by mensfeld** [comment]
Sorry for the noise but I want to trigger the CI at least 5 times in a row to make sure no other things go unnoticed. The fixes to specs explain why CI would crash but the operations would work as expected.

**Comment by mensfeld** [comment]
8/8 green after the last fixes. @thijsc feel free to take a look. This should conclude this PR. The stability on the Karafka side is unchanged (100% of things pass for producers and consumers).

At the moment ALL non-closeable objects that were leaking are closed correctly. Nothing that would not be finalized leaks to finalization or forks.

**Review by mensfeld — APPROVED** [review]

**Review by nijikon — APPROVED** [review]
@mensfeld great work on this! @thijsc you think that this can be merged now?

**Comment by nijikon** [comment]
@thijsc ping

**Pull request merged**